src/libmw/src/node/CoinsViewCache.cpp

```diff
@@ -0,0 +1,235 @@
+    m_pKernelMMR->Rewind(pHeader->GetNumKernels());
+    m_pOutputPMMR->Rewind(pHeader->GetNumTXOs());
+    SetBestHeader(pHeader);
+
+    // Sanity check to make sure rewind applied successfully
+    ValidateMMRs(pHeader);
+}
+
+mw::Block::Ptr CoinsViewCache::BuildNextBlock(const uint64_t height, const std::vector<mw::Transaction::CPtr>& transactions)
+{
+    LOG_TRACE_F("Building block with {} transactions", transactions.size());
+
+    auto pTransaction = Aggregation::Aggregate(transactions);
+
+    std::for_each(
+        pTransaction->GetKernels().cbegin(), pTransaction->GetKernels().cend(),
+        [this](const Kernel& kernel) { m_pKernelMMR->Add(kernel); }
+    );
+
+    std::for_each(
+        pTransaction->GetOutputs().cbegin(), pTransaction->GetOutputs().cend(),
+        [this, height](const Output& output) { AddUTXO(height, output); }
+    );
+
+    std::for_each(
+        pTransaction->GetInputs().cbegin(), pTransaction->GetInputs().cend(),
+        [this](const Input& input) { SpendUTXO(input.GetCommitment()); }
+    );
+
+    const uint64_t output_mmr_size = m_pOutputPMMR->GetNumLeaves();
+    const uint64_t kernel_mmr_size = m_pKernelMMR->GetNumLeaves();
+
+    mw::Hash output_root = m_pOutputPMMR->Root();
+    mw::Hash kernel_root = m_pKernelMMR->Root();
+    mw::Hash leafset_root = m_pLeafSet->Root();
+
+    BlindingFactor kernel_offset = pTransaction->GetKernelOffset();
+    if (GetBestHeader() != nullptr) {
+        kernel_offset = Crypto::AddBlindingFactors({
+            GetBestHeader()->GetKernelOffset(),
+            pTransaction->GetKernelOffset()
+        });
+    }
+
+    BlindingFactor owner_offset = pTransaction->GetOwnerOffset();
+
+    auto pHeader = std::make_shared<mw::Header>(
+        height,
+        std::move(output_root),
+        std::move(kernel_root),
+        std::move(leafset_root),
+        std::move(kernel_offset),
+        std::move(owner_offset),
+        output_mmr_size,
+        kernel_mmr_size
+    );
+
+    return std::make_shared<mw::Block>(pHeader, pTransaction->GetBody());
+}
+
+void CoinsViewCache::ValidateState() const
+{
+    StateValidator().Validate(*this);
+}
+
+bool CoinsViewCache::HasCoinInCache(const Commitment& commitment) const noexcept
+{
+    std::vector<CoinAction> actions = m_pUpdates->GetActions(commitment);
+    if (!actions.empty()) {
+        return actions.back().pUTXO != nullptr;
+    }
+
+    return false;
+}
+
+void CoinsViewCache::AddUTXO(const uint64_t header_height, const Output& output)
+{
+    mmr::LeafIndex leafIdx = m_pOutputPMMR->Add(output.ToOutputId());
+    m_pLeafSet->Add(leafIdx);
+
+    auto pUTXO = std::make_shared<UTXO>(header_height, std::move(leafIdx), output);
+
+    m_pUpdates->AddUTXO(pUTXO);
+}
+
+UTXO CoinsViewCache::SpendUTXO(const Commitment& commitment)
+{
+    std::vector<UTXO::CPtr> utxos = GetUTXOs(commitment);
+    if (utxos.empty() || !m_pLeafSet->Contains(utxos.back()->GetLeafIndex())) {
```

Comment:
Can !utxos.empty() && !m_pLeafSet->Contains(utxos.back()->GetLeafIndex()) ever be true? Maybe m_pLeafSet->Contains(utxos.back()->GetLeafIndex()) should be an assert only.